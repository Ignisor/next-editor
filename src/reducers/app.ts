const PUSH_SCENE = "app/push-scene"
const POP_SCENE = "app/pop-scene"
const REPLACE_SCENE = "app/replace-scene"

type PushScene = {
  type: typeof PUSH_SCENE
  payload: string
}

type PopScene = {
  type: typeof POP_SCENE
}

type ReplaceScene = {
  type: typeof REPLACE_SCENE
  payload: string
}

type Action = PushScene | PopScene | ReplaceScene

export function pushScene(nextScene: string): PushScene {
  return {
    type: PUSH_SCENE,
    payload: nextScene
  }
}

export function replaceScene(nextScene: string): ReplaceScene {
  return {
    type: REPLACE_SCENE,
    payload: nextScene
  }
}

export function popScene(nextScene: string): PopScene {
  return {
    type: POP_SCENE
  }
}

type LayoutMode = "editor" | "preview"
export type AppState = {
  sceneStack: string[]
  layoutMode: LayoutMode[]
}

const initialState: AppState = {
  sceneStack: ["editor"],
  layoutMode: ["editor", "preview"]
}

export function reducer(state: AppState = initialState, action: Action) {
  switch (action.type) {
    case PUSH_SCENE: {
      return { ...state, sceneStack: state.sceneStack.concat([action.payload]) }
    }
    case REPLACE_SCENE: {
      const length = state.sceneStack.length
      const popped = state.sceneStack.slice(0, length - 1)
      return { ...state, sceneStack: popped.concat([action.payload]) }
    }
    case POP_SCENE: {
      const length = state.sceneStack.length
      return { ...state, sceneStack: state.sceneStack.slice(0, length - 1) }
    }
    default: {
      return state
    }
  }
}
