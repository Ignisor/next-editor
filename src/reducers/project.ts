import { readDirectories } from "../domain/filesystem/queries/readDirectories"
import { createProject } from "../domain/git/commands/createProject"

type Project = {
  projectRoot: string
}

const LOAD_PROJECT_LIST = "project/load-project-list"

type LoadProjectList = {
  type: typeof LOAD_PROJECT_LIST
  payload: {
    projects: Project[]
  }
}

export async function loadProjectList(): Promise<LoadProjectList> {
  const projectNames = await readDirectories("/")
  return {
    type: LOAD_PROJECT_LIST,
    payload: {
      projects: projectNames.map(p => {
        return {
          projectRoot: p
        }
      })
    }
  }
}

export async function createNewProject(
  newProjectRoot: string
): Promise<LoadProjectList> {
  await createProject(newProjectRoot)
  return loadProjectList()
}

export type ProjectState = {
  projects: Project[]
}

const initialState: ProjectState = {
  projects: []
}

export type Action = LoadProjectList

export function reducer(state: ProjectState = initialState, action: Action) {
  switch (action.type) {
    case LOAD_PROJECT_LIST: {
      return { ...state, projects: action.payload.projects }
    }
    default: {
      return state
    }
  }
}
