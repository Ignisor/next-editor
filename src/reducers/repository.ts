import path from "path"
import { mkdir } from "../domain/filesystem/commands/mkdir"
import { unlink } from "../domain/filesystem/commands/unlink"
import { writeFile } from "../domain/filesystem/commands/writeFile"
import { addFile } from "../domain/git/commands/addFile"
import { checkoutBranch } from "../domain/git/commands/checkoutBranch"
import { commitChanges } from "../domain/git/commands/commitChanges"
import { removeFromGit } from "../domain/git/commands/removeFromGit"
import { getProjectGitStatus } from "../domain/git/queries/getProjectGitStatus"
import { listBranches } from "../domain/git/queries/listBranches"
import { GitRepositoryStatus } from "../domain/types"
import { RepositoryState } from "./repository"

type ThunkAction<A> = (
  dispatch: (a: A) => void | Promise<void>
) => void | Promise<void>

const j = path.join

const CHANGED = "repository:changed"
const GIT_STATUS_UPDATED = "repository:git-status-updated"

type Changed = {
  type: typeof CHANGED
  payload: {
    git?: boolean
    file?: boolean
    changedPath?: string
  }
}

type GitStatusUpdated = {
  type: typeof GIT_STATUS_UPDATED
  payload: GitRepositoryStatus
}

export function changed({
  git = true,
  file = true,
  changedPath = "/"
}: {
  git?: boolean
  file?: boolean
  changedPath?: string
} = {}): Changed {
  return {
    type: CHANGED,
    payload: {
      git,
      file,
      changedPath
    }
  }
}

export async function updateGitStatus(
  projectRoot: string
): Promise<GitStatusUpdated> {
  return {
    type: GIT_STATUS_UPDATED,
    payload: await getProjectGitStatus(projectRoot)
  }
}

export async function createFile(
  aPath: string,
  content: string = ""
): Promise<Changed> {
  await writeFile(aPath, content)
  const dirname = path.dirname(aPath)
  return changed({ changedPath: dirname })
}

export async function createBranch(projectRoot: string, newBranchName: string) {
  const branches = await listBranches(projectRoot)
  if (!branches.includes(newBranchName)) {
    await createBranch(projectRoot, newBranchName)
    console.log("create branch", newBranchName)
    return changed()
  } else {
    console.error(`Git: Creating branch existed: ${newBranchName}`)
  }
}

export function checkoutToOtherBranch(
  projectRoot: string,
  branchName: string
): ThunkAction<Changed> {
  return async dispatch => {
    const branches = await listBranches(projectRoot)
    if (branches.includes(branchName)) {
      await checkoutBranch(projectRoot, branchName)
      dispatch(changed())
    } else {
      console.error(`Git: Unknown branch: ${branchName}`)
    }
  }
}

export async function updateFile(
  aPath: string,
  content: string
): Promise<Changed> {
  await writeFile(aPath, content)
  const dirname = path.dirname(aPath)
  return changed({ changedPath: dirname })
}

export async function createDirectory(aPath: string) {
  await mkdir(aPath)
  const dirname = path.dirname(aPath)
  return changed({ changedPath: dirname })
}

export async function deleteFile(aPath: string) {
  await unlink(aPath)
  const dirname = path.dirname(aPath)
  return changed({ changedPath: dirname })
}

export async function addToStage(
  projectRoot: string,
  relpath: string
): Promise<Changed> {
  await addFile(projectRoot, relpath)
  const dirname = path.dirname(j(projectRoot, relpath))
  return changed({ changedPath: dirname })
}

export async function removeFileFromGit(
  projectRoot: string,
  relpath: string
): Promise<Changed> {
  await removeFromGit(projectRoot, relpath)
  const dirname = path.dirname(j(projectRoot, relpath))
  return changed({ changedPath: dirname })
}

export async function commitStagedChanges(
  projectRoot: string,
  message: string = "Update"
): Promise<Changed> {
  const author = {}
  const hash = await commitChanges(projectRoot, message)
  return changed({ changedPath: projectRoot })
}

type Action = GitStatusUpdated | Changed

export type RepositoryState = {
  gitRepositoryStatus: GitRepositoryStatus | null
  currentProjectRoot: string
  lastChangedPath: string
  lastChangedGithPath: string
  fsTouchCounter: number
  gitTouchCounter: number
}

const initialState: RepositoryState = {
  gitRepositoryStatus: null,
  currentProjectRoot: "/playground",
  gitTouchCounter: 0,
  lastChangedGithPath: "",
  lastChangedPath: "/playground",
  fsTouchCounter: 0
}

export function reducer(state: RepositoryState = initialState, action: Action) {
  switch (action.type) {
    case CHANGED: {
      return {
        ...state,
        fsTouchCounter: action.payload.file
          ? state.fsTouchCounter + 1
          : state.fsTouchCounter,
        gitTouchCounter: action.payload.git
          ? state.gitTouchCounter + 1
          : state.gitTouchCounter,
        lastChangedPath: action.payload.changedPath
      }
    }
    case GIT_STATUS_UPDATED: {
      return {
        ...state,
        gitRepositoryStatus: action.payload
      }
    }
    default: {
      return state
    }
  }
}
