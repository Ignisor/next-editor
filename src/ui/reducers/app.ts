import {
  ActionCreator,
  buildActionCreator,
  createReducer,
  Reducer
} from "hard-reducer"

const { createAction } = buildActionCreator({
  prefix: "app/"
})

export type AreaName = "editor" | "support"

export const setLayoutAreas: ActionCreator<{
  areas: AreaName[][]
}> = createAction("set-layout-mode")

export const pushScene: ActionCreator<{
  nextScene: string
}> = createAction("push-scene")

export const replaceScene: ActionCreator<{
  nextScene: string
}> = createAction("replace-scene")

export const popScene: ActionCreator<{}> = createAction("pop-scene")

export type Layout = {
  areas: AreaName[][]
  columns: string[]
  rows: string[]
}

export type AppState = {
  sceneStack: string[]
  mainLayout: Layout
}

const initialState: AppState = {
  sceneStack: ["main"],
  mainLayout: {
    columns: ["1fr", "1fr"],
    rows: ["1fr"],
    areas: [["editor", "support"]]
  }
}

export const reducer: Reducer<AppState> = createReducer(initialState)
  .case(setLayoutAreas, (state, payload) => {
    return {
      ...state,
      mainLayout: {
        ...state.mainLayout,
        areas: payload.areas
      }
    }
  })
  .case(pushScene, (state, payload) => {
    return {
      ...state,
      sceneStack: state.sceneStack.concat([payload.nextScene])
    }
  })
  .case(replaceScene, (state, payload) => {
    const length = state.sceneStack.length
    const popped = state.sceneStack.slice(0, length - 1)
    return { ...state, sceneStack: popped.concat([payload.nextScene]) }
  })
  .case(popScene, state => {
    const length = state.sceneStack.length
    return { ...state, sceneStack: state.sceneStack.slice(0, length - 1) }
  })
