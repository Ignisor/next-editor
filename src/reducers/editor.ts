import path from "path"
import { RootState } from "."
import { writeFile } from "../domain/filesystem/commands/writeFile"
import { readFile } from "../domain/filesystem/queries/readFile"
import { extToFileType } from "../lib/extToFileType"
import * as RepositoryActions from "./repository"
const CHANGE_VALUE = "editor/change-value"
const LOAD_FILE = "editor/load-file"

type ChangeValue = {
  type: typeof CHANGE_VALUE
  payload: {
    value: string
  }
}

type FILE_TYPES = "javascript" | "markdown" | "text"

type LoadFile = {
  type: typeof LOAD_FILE
  payload: {
    filePath: string
    fileType: FILE_TYPES
    value: string
  }
}

export type EditorState = {
  filePath: string | null
  fileType: FILE_TYPES
  loading: boolean
  value: string | null
}

export async function loadFile(filePath: string) {
  const fileContent = await readFile(filePath)

  return {
    type: LOAD_FILE,
    payload: {
      filePath,
      fileType: extToFileType(filePath),
      value: fileContent.toString()
    }
  }
}

export async function updateValue(filepath: string, value: string) {
  return async (
    dispatch: (a: Action | RepositoryActions.Action) => void,
    getState: () => RootState
  ) => {
    dispatch({
      type: CHANGE_VALUE,
      payload: {
        value
      }
    })
    await writeFile(filepath, value)
    // detect untracking changed
    const state = getState()
    if (state.repository.gitRepositoryStatus) {
      const root = state.repository.currentProjectRoot
      const relpath = path.relative(root, filepath)
      const { tracked, unstagedChanges } = state.repository.gitRepositoryStatus
      const alreadyChanged =
        unstagedChanges.findIndex(change => change.relpath === relpath) > -1
      if (!alreadyChanged && tracked.includes(relpath)) {
        dispatch(RepositoryActions.changed())
      }
    }
  }
}

const initialState: EditorState = {
  filePath: null,
  fileType: "text",
  loading: true,
  value: null
}

export type Action = ChangeValue | LoadFile

export function reducer(state: EditorState = initialState, action: Action) {
  switch (action.type) {
    case LOAD_FILE: {
      return { ...state, ...action.payload, loading: false }
    }
    case CHANGE_VALUE: {
      return { ...state, value: action.payload.value }
    }
    default: {
      return state
    }
  }
}
