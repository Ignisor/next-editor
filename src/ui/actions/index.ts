import * as app from "../reducers/app"
import * as buffer from "../reducers/buffer"
import * as config from "../reducers/config"
import * as git from "../reducers/git"
import * as project from "../reducers/project"
import * as repository from "../reducers/repository"
import * as editor from "./editorActions"

export default { app, buffer, project, repository, git, config, editor }
