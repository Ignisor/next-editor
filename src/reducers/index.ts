import { combineReducers } from "redux"
import { EditorState, reducer as editor } from "./editor"
import { ProjectState, reducer as project } from "./project"
import { reducer as repository, RepositoryState } from "./repository"

export type RootState = {
  repository: RepositoryState
  project: ProjectState
  editor: EditorState
}

export const rootReducer: (
  state: RootState,
  action: any
) => RootState = combineReducers({
  project,
  editor,
  repository
} as any)
