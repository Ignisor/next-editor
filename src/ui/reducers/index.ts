import { connect } from "react-redux"
import { InferableComponentEnhancerWithProps } from "recompose"
import { combineReducers, compose } from "redux"
import * as app from "./app"
import * as config from "./config"
import * as editor from "./editor"
import * as git from "./git"
import * as project from "./project"
import * as repository from "./repository"

const allActions = { app, editor, project, repository, git, config }

export type RootState = {
  app: app.AppState
  repository: repository.RepositoryState
  project: project.ProjectState
  editor: editor.EditorState
  git: git.GitState
  config: config.ConfigState
}

export type AllAction = typeof allActions

export const rootReducer: (
  state: RootState,
  action: any
) => RootState = combineReducers({
  app: app.reducer,
  project: project.reducer,
  editor: editor.reducer,
  repository: repository.reducer,
  git: git.reducer,
  config: config.reducer
} as any)

export const connector = <
  OwnProps extends {} = {},
  Connected extends {} = {},
  BoundAction extends {} = {}
>(
  stateSelector: (state: RootState, ownProps: OwnProps) => Connected,
  actionSelector: (actions: AllAction) => BoundAction,
  // ...hocs: Array<(props: Connected & BoundAction) => any>
  ...hocs: any[]
): InferableComponentEnhancerWithProps<Connected & BoundAction, OwnProps> => {
  return (compose as any)(
    connect(
      stateSelector,
      actionSelector(allActions)
    ),
    ...hocs
  )
}
