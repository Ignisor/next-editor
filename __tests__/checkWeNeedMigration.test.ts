import { rootReducer } from "../src/ui/reducers"

// This is special test for migration checker
test("Check we need migration", () => {
  const initialState = rootReducer(undefined as any, { type: "init" })
  expect(initialState).toMatchSnapshot()
})
