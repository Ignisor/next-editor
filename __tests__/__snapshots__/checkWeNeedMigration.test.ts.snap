// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check we need migration 1`] = `
Object {
  "app": Object {
    "mainLayout": Object {
      "areas": Array [
        Array [
          "menu",
          "editor",
          "support",
        ],
      ],
      "columns": Array [
        "250px",
        "1fr",
        "1fr",
      ],
      "rows": Array [
        "1fr",
      ],
    },
    "sceneStack": Array [
      "main",
    ],
  },
  "buffer": Object {
    "autosave": true,
    "changed": false,
    "filepath": "",
    "filetype": "text",
    "lastSavedValue": "",
    "reloadCounter": 0,
    "value": "",
  },
  "config": Object {
    "committerEmail": "",
    "committerName": "",
    "doneTutorial": false,
    "editorFontFamily": "Inconsolata, monospace",
    "editorFontScale": 1,
    "editorSpellCheck": false,
    "githubApiToken": "",
    "githubProxy": "https://cors-buster-zashozaqfk.now.sh/github.com/",
    "isFirstVisit": true,
    "theme": "dark",
  },
  "git": Object {
    "branches": Array [],
    "currentBranch": null,
    "history": Array [],
    "loaded": false,
    "projectRoot": null,
    "remoteBranches": Array [],
    "remotes": Array [],
    "staging": null,
    "stagingLoading": true,
    "type": "loading",
  },
  "project": Object {
    "projectRoot": "/",
    "projects": Array [],
  },
  "repository": Object {
    "currentProjectRoot": "/playground",
    "dirCreatingDir": null,
    "fileCreatingDir": null,
    "renamingFilepath": null,
    "touchCounter": 0,
  },
}
`;
